### PR TITLE
[Auditbeat] System module: Update and re-enable package dataset

### PR DESCRIPTION
**Pull request by cwurm**
Re-enables the disabled `package` dataset and brings it up to date with the other, soon-to-be released datasets.

High-level changes:

- Renamed to `package` (singular)
- Scheduled state reporting based on `state.period` and `package.state.period`
- Common fields: `event.kind`, `event.action`, `event.id`, `message`
- Save/Restore package information to disk

Unfortunately, the changes to `package.go` are extensive enough that the Github diff view presents it as a new file. A lot of lines have indeed changed, though none of the concepts are net new, they either exist in the other datasets or in the disabled implementation of the dataset.

Follow-ups, already listed in https://github.com/elastic/beats/issues/10103:

- [ ] Improve Homebrew package collection: parse `INSTALL_RECEIPT.json`
- [ ] RPM support (https://github.com/elastic/beats/pull/9092)
- [ ] Dashboard
- [ ] More and better tests

### CODE REVIEW

File: x-pack/auditbeat/module/system/package/package.go [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+}
+
+func (pkg Package) toMapStr() common.MapStr {
+	mapstr := common.MapStr{
+		"name":        pkg.Name,
+		"version":     pkg.Version,
+		"installtime": pkg.InstallTime,
+	}
+
+	if pkg.Arch != "" {
```
Comment (webmat) [inline x-pack/auditbeat/module/system/package/package.go]: Nit: Couldn't you use `putIfNotEmpty` here as well?

File: x-pack/auditbeat/module/system/package/package.go [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+
+	bucket, err := datastore.OpenBucket(bucketName)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to open persistent datastore")
+	}
+
+	ms := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+		log:           logp.NewLogger(metricsetName),
```
Comment (webmat) [inline x-pack/auditbeat/module/system/package/package.go]: A what? A "metricset"? ;-)

File: x-pack/auditbeat/module/system/packages/packages.go [hunk 3]
```diff
@@ -1,338 +0,0 @@
-// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
-// or more contributor license agreements. Licensed under the Elastic License;
-// you may not use this file except in compliance with the Elastic License.
-
-package packages
```
Comment (tsg) [inline x-pack/auditbeat/module/system/packages/packages.go]: How come this file was deleted? Does it exists somewhere else now?

Comment (tsg) [reply x-pack/auditbeat/module/system/packages/packages.go]: Oops, nevermind, saw it now. It was renamed to `package/package.go`.

File: x-pack/auditbeat/module/system/package/package.go (hunk 1 above)
Comment (cwurm) [reply x-pack/auditbeat/module/system/package/package.go]: I could have indeed, and I briefly thought about it but shied away from copying the function. Looking at more of this type of code throughout the Beats codebase I think `putIfNotEmpty` should be added to `common.MapStr` so we can use it everywhere. I'll make a note to propose adding it when I have a minute.

File: x-pack/auditbeat/module/system/package/package.go (hunk 2 above)
Comment (cwurm) [reply x-pack/auditbeat/module/system/package/package.go]: Indeed :) We might never be able to get it out of the whole codebase. For now at least everything that is user facing calls it "dataset".

File: x-pack/auditbeat/module/system/package/package.go [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+	if !ms.lastState.IsZero() {
+		ms.log.Debugf("Last state was sent at %v. Next state update by %v.", ms.lastState, ms.lastState.Add(ms.config.effectiveStatePeriod()))
+	} else {
+		ms.log.Debug("No state timestamp found")
+	}
+
+	// Load from disk: Packages
+	packages, err := ms.restorePackagesFromDisk()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to restore packages from disk")
```
Comment (tsg) [inline x-pack/auditbeat/module/system/package/package.go]: I think that returning an error here is going to cause Auditbeat to stop, right? Not sure if we discussed this before, but I wonder if recovering from this condition is not something that we want, to protect from corrupted gob.

We could print a stacktrace to the logs with a clear "please report this" message, but still continue the process.

I'm not convinced that we should recovered from the error, but I think it's worth discussing.

Comment (cwurm) [reply x-pack/auditbeat/module/system/package/package.go]: > I think that returning an error here is going to cause Auditbeat to stop, right?

Yes. If the `beat.db` file is corrupted, the user will have to delete it to start.

My first instinct is to be conservative and surface the error. Something unexpected happened - maybe innocent, maybe not, we don't know - and the user should address it before continuing.

One alternative would be to print a warning but continue, e.g. treating the file as non existent which would trigger a whole state update to be sent. But that would miss any changes made while Auditbeat was down.

I think my opinion at this time is that unless we have a somewhat common case in which this error can be triggered and there is a good reason to not stop in that case I would be in favor of stopping as we do now.

File: x-pack/auditbeat/module/system/package/package.go [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+				Name:        packageDir.Name(),
+				Version:     version.Name(),
+				InstallTime: version.ModTime(),
+			}
+
+			// read formula
+			formulaPath := path.Join(homebrewCellarPath, pkg.Name, pkg.Version, ".brew", pkg.Name+".rb")
+			file, err := os.Open(formulaPath)
+			if err != nil {
+				//fmt.Printf("WARNING: Can't get formula for package %s-%s\n", pkg.Name, pkg.Version)
```
Comment (tsg) [inline x-pack/auditbeat/module/system/package/package.go]: Log a warning or debug message?

File: x-pack/auditbeat/module/system/package/package.go (hunk 4 above)
Comment (tsg) [reply x-pack/auditbeat/module/system/package/package.go]: Ok, I agree. Let's start with stopping Auditbeat in this case.

My concern is that a break in compatibility in `beat.db` is going to escalate itself to blocker bugs (Auditbeat doesn't start at all). In order to protect against this, we should have tests with some `beat.db` golden files from older versions.

File: x-pack/auditbeat/auditbeat.reference.yml [hunk 6]
```diff
@@ -117,6 +117,7 @@ auditbeat.modules:
 - module: system
   datasets:
     - host    # General host information, e.g. uptime, IPs
+    - package # Installed and removed packages
```
Comment (andrewkroh) [inline x-pack/auditbeat/auditbeat.reference.yml]: Installed / updated / removed packages?

File: x-pack/auditbeat/module/system/_meta/config.yml.tmpl [hunk 7]
```diff
@@ -34,8 +34,4 @@
   # detect any changes.
```
Comment (andrewkroh) [inline x-pack/auditbeat/module/system/_meta/config.yml.tmpl]: Is there a `package.state.period`?

File: x-pack/auditbeat/module/system/package/package.go (hunk 5 above)
Comment (cwurm) [reply x-pack/auditbeat/module/system/package/package.go]: Ah yes, good point. We don't have a logger in hand there at this time unfortunately. But I'll open a follow-up PR to improve Homebrew package collection that is going to address this as well. I have the code mostly ready.

**Comment by cwurm** [comment]
jenkins, test this

**Pull request merged**